internal/evaluation_target/application_snapshot_image/application_snapshot_image.go

```diff
@@ -86,6 +92,18 @@ func NewApplicationSnapshotImage(ctx context.Context, url string, p policy.Polic
 		return nil, err
 	}
 
+	// TODO: At some point, the SignatureValidator is no longer a special case, and it can
+	// be enabled or disabled via the source groups in the PolicyConfiguration.
+	signatureValidator, err := validator.NewImageValidator(
+		signature.SignatureValidatorName, validator.Options{Policy: p})
+	if err != nil {
+		return nil, err
+	}
+	a.Validators = append(a.Validators, signatureValidator)
+
+	// TODO: Populate a.Validators from sourceGroup. Initially, assume that they all
+	// want ProvenanceV0_2Validator.
```

Comment:
We should use something like `validator.Validate(Context, Artifact)` which in turn would loop over and find all applicable validators. Bit of code here should not care about specific validators at all.

Reply:
I agree that this piece of code shouldn't care about which validator is being executed. The long-term idea here is that the set of validators that are included is completely configuration driven. For example, I'm picturing the policy config (contract) will be expressed as something like this:

```yaml
validators:  # This bit instead of `sources`
  - name: image-signature
    policy: ...
    data: ...
 
  # Apply policies to the config resource associated with the image. For example, to validate
  # that the image contains all the required labels an enterprise expects.
  - name: image-config
    policy: ...
    data: ...

    # This should eventually evolve into something like `in-toto-slsa-provenance` which can
    # handle any supported version of SLSA Provenance. We may also want to introduce a
    # a new validator `in-toto-slsa-provenance-v1.0` for those who care about that particular
    # version
  - name: in-toto-slsa-provenance-v0.2
    policy: ...
    data: ...

  - name: in-toto-cyclonedx-sbom
    policy: ...
    data: ...
```

Once we get to some version of that, we can loop over the validators, get each one by name (without having to care about its implementation), and simply execute them. (I can see this even leveraging a plugin system in the future if the need arises.) We're not there yet, so this is now hard-coded to add the validators that match the current state of what "validate image" does.

The reason I like this approach is that it gives policy authors a lot of flexibility while making it obvious what is actually being validated. Long-term, I don't think the ec-cli should have the logic in determining which validators to be executed because that will be difficult to get it right for multiple use cases without cluttering the policy config.

@zregvart, wdyt of this reasoning?

Reply:
I see that this could work, though I don't think it is ambitious enough.

I'd go to the other extreme, I don't want to do it in configuration I want to do it via policy rules. Given the configuration it only applies to images and certain version of SLSA provenance, I would like to be in a place where I can say my contract is:

- artifact needs to be signed
- provenance needs to comply to SLSA vX

Where the internal details of the rule is what it means that an artifact is signed and by whome and that it conforms to a SLSA level. Basically having the configuration (contract) just mention what policy rules need to apply and if needed parameters for those.